### PR TITLE
docs(core/service-worker): document `isStable` traps

### PR DESCRIPTION
**Pull request by cexbrayat**
The docs don't mention that the app will never be stable if a `setInterval` is running somewhere, and that it will prevent the servcie worker to be registered too.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the new behavior?

Add an extra note about `setInterval` and `isStable` after #28020 fix.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by gkalpak — COMMENTED** [review]
I think it might be better adding such a note in the [isStable API docs][1] and then link to it from here.
It may easily catch people off-guard with `isStable`.

[1]: https://next.angular.io/api/core/ApplicationRef#isStable

**Comment by cexbrayat** [comment]
@gkalpak I added a few examples on `isStable` about the things that caught us when trying to use it (in a first commit). I added a sentence in the service worker documentation (the commit you reviewed, which is now the second commit). I was hoping that the link would be added automatically on `isStable` by the auto-link process, but it isn't... Do I need to make an explicit link?

**Comment by gkalpak** [comment]
Thx, @cexbrayat 👍
The auto-linking process cannot disambiguate `isStable`, because there are [several of them][1]. So, yes, you need to add an explicit link (e.g. `{@link ApplicationRef#isStable}` iirc).

[1]: https://angular.io/?search=isStable

**Review by gkalpak — CHANGES_REQUESTED** [review]

**Comment by cexbrayat** [comment]
@gkalpak Thanks for the review. 

I pushed two fixup commits to make it easier to review the changes based on your feedback. Let me know if it's OK and if I need to squash them (I don't understand if that's automatically done by your process or not, as a commit with `fixup!` ended on master yesterday ^^).

**Comment by gkalpak** [comment]
Taking a look now, thx. Fixup commits are fine.
(The reason one ended up in master yesterday is that it was fixing up a non-existent commit (see [PR commits][1]). We should add a check in the merge scripts.)

[1]: https://github.com/angular/angular/pull/28066/commits

File: aio/content/guide/service-worker-communications.md [hunk 1]
```diff
@@ -50,6 +50,9 @@ script will never be registered with the browser.
 You can avoid that by waiting for the app to stabilize first, before starting to poll for updates
 (as shown in the example above).
 
+Note that this is true for any kind of polling done by your application!
+Check the {@link ApplicationRef#isStable} documentation for more information.
```
Comment (gkalpak) [inline aio/content/guide/service-worker-communications.md]: My bad; actually this results in a link that looks like [ApplicationRef](https://angular.io/api/core/ApplicationRef#isStable). I expected it to be [isStable](https://angular.io/api/core/ApplicationRef#isStable).
You can set the text explicitly with `{@link ApplicationRef#isStable isStable}`.

File: packages/core/src/application_ref.ts [hunk 2]
```diff
@@ -373,6 +373,89 @@ export class ApplicationRef {
...
+   *      filter(stable => stable)
+   *   ).subscribe(() => console.log('App is stable now');
+   * }
+   * ```
+   * In this example, `isStable` will never emit `true`,
+   * and the trace "App is stable now" will never get logged.
+   *
+   * If you want to execute something when the app is stable,
+   * you'll have to delay your polling process,
+   * or wait for the application to be stable before starting it.
```
Comment (gkalpak) [inline packages/core/src/application_ref.ts]: This sounds confusing (_"delay polling"_ or _"wait for stable"_). How about:

_"you have to delay your polling process until the application is stable"_
or
_"you have to wait for the application to be stable before starting your polling process"_

File: packages/core/src/application_ref.ts [hunk 3]
```diff
@@ -373,6 +373,89 @@ export class ApplicationRef {
...
+   * Note two important points, demonstrated in the examples below:
+   * - the application will never be stable if you start any kind
+   * of asynchronous task when the application starts
+   * (for example for a polling process, started with a `setInterval`, a `setTimeout`
+   * or using RxJS operators like `interval`);
+   * - the `isStable` Observable runs outside of the Angular zone.
+   *
+   * Let's imagine that you start a recurrent task
+   * (here incrementing a counter, using RxJS `interval`),
+   * and then subscribe to `isStable`.
```
Comment (gkalpak) [inline packages/core/src/application_ref.ts]: The _"then"_ makes it sound as if order makes a difference.
Consider changing to _"...and at the same time..."_. (And potentially moving `interval(1000).subscribe(...)` after `isStable.subscribe(...)` to make it clear that order does not matter.)

File: packages/core/src/application_ref.ts [hunk 4]
```diff
@@ -373,6 +373,89 @@ export class ApplicationRef {
 
   /**
    * Returns an Observable that indicates when the application is stable or unstable.
+   *
+   * Note two important points, demonstrated in the examples below:
+   * - the application will never be stable if you start any kind
+   * of asynchronous task when the application starts
```
Comment (gkalpak) [inline packages/core/src/application_ref.ts]: It is important that the async task is recurrent. A one-off async task will just delay stabiliation, which is generally not a problem.

**Review by gkalpak — APPROVED** [review]
Left a couple of minor comments. LGTM otherwise 👍

**Comment by cexbrayat** [comment]
@gkalpak 👍 two new commits to review for you

**Review by gkalpak — APPROVED** [review]
👌

File: packages/core/src/application_ref.ts [hunk 5]
```diff
@@ -373,6 +373,89 @@ export class ApplicationRef {
 
   /**
    * Returns an Observable that indicates when the application is stable or unstable.
+   *
```
Comment (IgorMinar) [inline packages/core/src/application_ref.ts]: This content is great, but if you take a look at the preview: https://pr28102-c0a19d7.ngbuilds.io/api/core/ApplicationRef#properties you’ll see that that’s the wrong place for that content.

if @jbogart took a look at this PR she’d tell us that that content belongs to usage notes section for ApplicationRef with a note in the “isStable” jsdoc to check out the usage notes for caveats to be aware of when using that API.

can you please make such changes? thanks!

**Review by IgorMinar — DISMISSED** [review]

File: packages/core/src/application_ref.ts (hunk 5 above)
Comment (jenniferfell) [reply packages/core/src/application_ref.ts]: And I'd say: Remove the exclamation point. Let's stop yelling at readers. :-)

Comment (cexbrayat) [reply packages/core/src/application_ref.ts]: @IgorMinar I moved the examples to usage notes and added a link in `isStable`: let me know if that's better https://pr28102-3fbf6c7.ngbuilds.io/api/core/ApplicationRef

@jenniferfell Exclamation point removed 🙂

File: packages/core/src/application_ref.ts [hunk 6]
```diff
@@ -348,6 +348,93 @@ function optionsReducer<T extends Object>(dst: any, objs: T | T[]): T {
 /**
  * A reference to an Angular application running on a page.
  *
+ * @usageNotes
+ *
+ * ### isStable examples and caveats
```
Comment (gkalpak) [inline packages/core/src/application_ref.ts]: You can link directly to this section (using `#isstable-examples-and-caveats`), but even better (and future-proof) is to create a shorter anchor using `{@a name-of-anchor}` and use that for linking.
(Reference: [Authors Style Guide > Anchors][1])

[1]: https://angular.io/guide/docs-style-guide#ugly-long-section-header-anchors

Comment (cexbrayat) [reply packages/core/src/application_ref.ts]: @gkalpak Done 👍

**Comment by cexbrayat** [comment]
@gkalpak I forgot to ping you when I fixed your last feedback. Is there anything to do to get this merged?

**Comment by gkalpak** [comment]
merge-assistance: Requires g3sync.
merge-assistance: Using my global approval, since this is a docs-only change.

**Comment by matsko** [comment]
http://test/OCL:232707992:BASE:232708128:1549480177938:ecd9d18a

**Pull request closed without merging**